### PR TITLE
feat: REPLAT-4861 Integrate tex-layout and use for dropcaps

### PR DESCRIPTION
**Pull request by jmars**
Implements drop caps using the new text-layout utility library, no native dependencies are used so now dropcaps also work on IOS. This also lays the ground work for using the library for inline images and other text flowing components.

screenshot_20190129_130008

### CODE REVIEW

File: packages/article-paragraph/__tests__/android/__snapshots__/article-paragraph.android.test.js.snap [hunk 1]
```diff
@@ -17,18 +17,107 @@ exports[`3. paragraph with a drop cap 1`] = `
...
-  >
-    his being Black History 
-  </Text>
-  <Text
-    selectable={true}
-  >
-    Month, last week was Politicians In Search Of An Eye-Catching Race-Related Policy Week.
+    This being Black History Month, last week was Politicians In Search Of An Eye-Catching Race-Related Policy Week.
   </Text>
+  <View>
```
Comment (tuncaulubilge) [inline packages/article-paragraph/__tests__/android/__snapshots__/article-paragraph.android.test.js.snap]: This looks a bit odd

File: package.json [hunk 2]
```diff
@@ -6,7 +6,7 @@
   "main": "index.js",
   "engines": {
     "node": ">=8.10.0 <10",
-    "yarn": "1.12.3"
+    "yarn": ">=1.12.3"
```
Comment (tuncaulubilge) [inline package.json]: Let's keep this as a fixed version, as yarn had some inconsistencies in the previous upgrades a few times. We should do a separate pr if we want to bump to latest yarn

Comment (jmars) [reply package.json]: will revert this it was not meant to be committed, I just have yarn 1.13 locally

File: packages/article-paragraph/__tests__/android/__snapshots__/article-paragraph.android.test.js.snap (hunk 1 above)
Comment (jmars) [reply packages/article-paragraph/__tests__/android/__snapshots__/article-paragraph.android.test.js.snap]: Looks like it is taking a snapshot before measurement finishes

Comment (jimhunty) [reply packages/article-paragraph/__tests__/android/__snapshots__/article-paragraph.android.test.js.snap]: I think a `sleep` is used elsewhere in tests. Feels unpredictable to me though.

File: package.json (hunk 2 above)
Comment (jimhunty) [reply package.json]: `npx yarn@1.12.3`

File: packages/utils/src/text-layout/index.js [hunk 3]
```diff
@@ -0,0 +1,178 @@
...
+  };
+};
+
+export const measureInline = elements => {
+  const els = [];
+  const promises = elements.map(
+    (el, i) =>
+      new Promise(resolve => {
+        els.push({
+          ...el,
```
Comment (nathggns) [inline packages/utils/src/text-layout/index.js]: Have you considered using `React.cloneElement` instead? Just a more standardised way of cloning a component with updated props.

File: packages/utils/src/text-layout/index.js [hunk 4]
```diff
@@ -0,0 +1,178 @@
...
+
+export const measureText = texts => {
+  const els = [];
+  const promises = texts.map(
+    (text, i) =>
+      new Promise(resolve => {
+        const key = String(i);
+        els.push(
+          <Text
+            {...text.props}
```
Comment (nathggns) [inline packages/utils/src/text-layout/index.js]: Have you considered using `React.cloneElement` instead? Just a more standardised way of cloning a component with updated props.

File: packages/utils/src/text-layout/index.js [hunk 5]
```diff
@@ -0,0 +1,178 @@
...
+    {children}
+  </View>
+);
+
+export const measureText = texts => {
+  const els = [];
+  const promises = texts.map(
+    (text, i) =>
+      new Promise(resolve => {
+        const key = String(i);
```
Comment (nathggns) [inline packages/utils/src/text-layout/index.js]: As discussed, don't need to cast these to a string, just disable the lint rule for this file.

File: packages/utils/src/text-layout/formatter.js [hunk 6]
```diff
@@ -0,0 +1,36 @@
...
+
+export const defaults = {
+  space: {
+    shrink: 9,
+    stretch: 6,
+    width: 3
+  }
+};
+
+export default words => ({
```
Comment (nathggns) [inline packages/utils/src/text-layout/formatter.js]: Probably worth naming this function and then `export default name` or whatever just to make call stacks and debugging more clear

File: packages/article-paragraph/__tests__/shared.native.js [hunk 7]
```diff
@@ -40,6 +40,11 @@ export default () => {
         );
 
         await delay(0);
+
+        callAllLayouts(testInstance);
+
+        await delay(0);
```
Comment (nathggns) [inline packages/article-paragraph/__tests__/shared.native.js]: Might be worth including these delays as part of the callAllLayouts function if they're necessary to be able to examine the results of it.

File: packages/article-paragraph/src/drop-cap.js [hunk 8]
```diff
@@ -100,43 +94,46 @@ class DropCapParagraph extends Component {
...
-            >
-              {slicePoint > 0 ? text.slice(slicePoint).trim() : null}
-            </Text>
           </View>
         )}
-      </ResponsiveContext.Consumer>
-    );
+      </InlineElement>,
+      <Text selectable style={stylesThemedAndScaled.articleTextElement}>
+        {text.trim()}
```
Comment (nathggns) [inline packages/article-paragraph/src/drop-cap.js]: As discussed, this is an implementation detail and should be taken care of in the layout algorithm.

File: packages/utils/src/text-layout/index.js [hunk 9]
```diff
@@ -0,0 +1,178 @@
...
+
+  const { elements: wordElements, results: wordsResults } = measureText(
+    splitEls
+  );
+  const {
+    elements: inlineElements,
+    results: inlineSizesResults
+  } = measureInline(inlines.map(el => el.props.children({})));
+
+  const results = new Promise(async resolve => {
```
Comment (nathggns) [inline packages/utils/src/text-layout/index.js]: This is a bit weird… don't really need the new Promise constructor for this.

File: packages/utils/src/text-layout/index.js [hunk 10]
```diff
@@ -0,0 +1,178 @@
...
+        } else {
+          result.push(node);
+        }
+      }
+      tolerance += 1;
+    }
+    const height = paragraph[0].height * lines;
+    return [result, height];
+  }
+  return undefined;
```
Comment (nathggns) [inline packages/utils/src/text-layout/index.js]: This is unneeded.

File: packages/utils/src/screen.base.js [hunk 11]
```diff
@@ -21,5 +20,4 @@ export const normaliseWidth = width => {
   return nWidth || acceptedWidths[acceptedWidths.length - 1];
 };
 
-export const screenWidth = isTablet =>
-  isTablet ? tabletWidth - tabletRowPadding : Dimensions.get("window").width;
+export const screenWidth = () => Dimensions.get("window").width;
```
Comment (jimhunty) [inline packages/utils/src/screen.base.js]: This change will cause a lot of problems elsewhere. Has the full scope of this change been looked at from a styling perspective?

File: packages/utils/src/text-layout/align.js [hunk 12]
```diff
@@ -0,0 +1,159 @@
+import React from "react";
```
Comment (jimhunty) [inline packages/utils/src/text-layout/align.js]: `text-layout` should be a package in itself in my opinion.

File: packages/article-paragraph/src/drop-cap.js [hunk 13]
```diff
@@ -1,94 +1,88 @@
...
+      text !== pText ||
+      dropCap !== pDropCap
+    ) {
+      // eslint-disable-next-line react/no-did-update-set-state
+      this.setState(
+        {
+          content: [],
+          height: null,
+          needsLayout: true,
+          screenWidth: Dimensions.get("window").width
```
Comment (jimhunty) [inline packages/article-paragraph/src/drop-cap.js]: Changes to screen width utility function but then not used here?

File: packages/utils/src/screen.base.js (hunk 11 above)
Comment (jmars) [reply packages/utils/src/screen.base.js]: Was a bad rebase, fixed now
